### PR TITLE
Merge bitcoin#13151: net: Serve blocks directly from disk when possible

### PR DESCRIPTION
**Pull request by PastaPastaPasta**

### CODE REVIEW

File: src/validation.cpp [hunk 1]
```diff
@@ -1070,6 +1070,52 @@ bool ReadBlockFromDisk(CBlock& block, const CBlockIndex* pindex, const Consensus
...
+    CAutoFile filein(OpenBlockFile(hpos, true), SER_DISK, CLIENT_VERSION);
+    if (filein.IsNull()) {
+        return error("%s: OpenBlockFile failed for %s", __func__, pos.ToString());
+    }
+
+    try {
+        CMessageHeader::MessageStartChars blk_start;
+        unsigned int blk_size;
+
+        filein >> blk_start >> blk_size;
```
Comment (PastaPastaPasta) [inline src/validation.cpp:1086]: compilation fails here, not sure if I'm missing a previous backport or what. 
Unable to call Unserialize for type obj

Comment (UdjinM6) [reply src/validation.cpp:1086]: `ReadRawBlockFromDisk` introduced here is only used for `MSG_WITNESS_BLOCK` which we don't have. I think it safe to do a partial backport and drop `ReadRawBlockFromDisk` functions.

**Comment by PastaPastaPasta** [comment]
Dropped those two functions, so I think it'd be safe to see we won't see the large speed-ups that upstream did?

**Comment by PastaPastaPasta** [comment]
rebased on develop

File: src/net_processing.cpp [hunk 2]
```diff
@@ -1451,8 +1452,10 @@ void static ProcessGetBlockData(CNode* pfrom, const Consensus::Params& consensus
                 assert(!"cannot load block from disk");
             pblock = pblockRead;
         }
-        if (inv.type == MSG_BLOCK)
-            connman->PushMessage(pfrom, msgMaker.Make(NetMsgType::BLOCK, *pblock));
+        if (pblock) {
+            if (inv.type == MSG_BLOCK)
+                connman->PushMessage(pfrom, msgMaker.Make(NetMsgType::BLOCK, *pblock));
+        }
```
Comment (UdjinM6) [inline src/net_processing.cpp:1458]: This is incorrect, the whole `if (inv.type == MSG_BLOCK) ... else if (inv.type == MSG_FILTERED_BLOCK) ... else if (inv.type == MSG_CMPCT_BLOCK) ...` thing should be wrapped into `if (pblock)` instead.

Comment (PastaPastaPasta) [reply src/net_processing.cpp]: bb6c6597651a97ea76fc73f9a6f0f68714e7b55f

**Comment by UdjinM6** [comment]
Needs rebase

**Comment by PastaPastaPasta** [comment]
rebased on develop

**Comment by PastaPastaPasta** [comment]
rebased again to hopefully resolve tests

**Pull request merged**